cmd/pj-rehearse/main.go

```diff
@@ -572,3 +571,37 @@ func ensureImageStreamTags(ctx context.Context, client ctrlruntimeclient.Client,
 
 	return g.Wait()
 }
+
+// determineSubsetToRehearse determines in a sophisticated way which subset of jobs should be chosen to be rehearsed.
+// First, it will create a list of the jobs mapped by the source type and calculates the maximum allowed jobs for each
+// source type. If there are jobs from a specific source type that are under the max allowed number, it will fill the gab
+// from the other not chosen jobs until it reaches the rehearsal limit.
+func determineSubsetToRehearse(presubmitsToRehearse []*prowconfig.Presubmit, rehearsalLimit int) []*prowconfig.Presubmit {
+	if len(presubmitsToRehearse) <= rehearsalLimit {
+		return presubmitsToRehearse
+	}
+
+	presubmitsBySourceType := make(map[config.SourceType][]*prowconfig.Presubmit)
+	for _, p := range presubmitsToRehearse {
+		sourceType := config.GetSourceType(p.Labels)
+		presubmitsBySourceType[sourceType] = append(presubmitsBySourceType[sourceType], p)
+	}
+
+	maxJobsPerSourceType := rehearsalLimit / len(presubmitsBySourceType)
+	toRehearse := []*prowconfig.Presubmit{}
+	dropped := []*prowconfig.Presubmit{}
+
+	for _, jobs := range presubmitsBySourceType {
+		if len(jobs) > maxJobsPerSourceType {
+			toRehearse = append(toRehearse, jobs[:maxJobsPerSourceType]...)
+			dropped = append(dropped, jobs[maxJobsPerSourceType:]...)
+		}
+	}
+
+	if len(toRehearse) <= rehearsalLimit {
```

Comment:
Can you add a comment here that this might happen due to the rounding inherent in integer division? Also, can we get a test for when you need to add more?

Reply:
Comment added as well as a new test case for this.